### PR TITLE
Refactor DGLKE

### PR DESCRIPTION
**Pull request by menjarleev**
*Issue #, if available:*
- Currently only test TransE_l1 with wn18. Other models need to be tested in the future.

*Description of changes:*

- refactor entire DGLKE package to enable further extension.
- Package can are divided into: Model, Dataset, Dataloader.
- Dataset download benchmark dataset(FB15k) from web and split it into train, eval, test datasets. These datasets can be further partitioned into chunks to support single machine multi-GPU training.
- Dataloader sample data from partition dataset and feed them to model.
- A model consists of an encoder and (possibly multiple) decoder. The encoder fetch samples from dataloader, and output embeddings of graph based on the sliced indices. The decoder use embeddings from encoder to produce error and update parameters of model or generate evaluation results.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by zheng-da** [comment]
please PR to another branch instead of overwriting v0.2.

**Comment by zheng-da** [comment]
can you PR to the `ng` branch? `v0.2` branch is used for releasing code.

File: python/refactor/main.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
+from .utils.argparser import TrainArgParser
+from .utils.misc import set_seed, prepare_args
+from .utils.logging import Logger
+from .utils.data.dataset import get_dataset, TrainDataset, TestDataset, ValidDataset
```
Comment (classicsong) [inline python/refactor/main.py:4]: Make data as the first level module.

File: python/refactor/main.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+                          format_str=args.format,
+                          delimiter=args.delimiter,
+                          files=args.data_files,
+                          has_edge_importance=args.has_edge_importance)
+    train_dataset, eval_dataset, test_dataset = None, None, None
+    # create training dataset needed parameters
+    train_dataset = TrainDataset(dataset, args)
+    g = train_dataset.g
+    args.strict_rel_part = args.mix_cpu_gpu and (train_dataset.cross_part is False)
+    args.soft_rel_part = args.mix_cpu_gpu and args.rel_part and train_dataset.cross_part is True
```
Comment (classicsong) [inline python/refactor/main.py:39]: We can forget soft_rel_part

File: python/refactor/main.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+
+def create_dataloader(args):
+    train_dataloader, eval_dataloader = None, None
+    if args.model == 'KGE':
+        train_metadata = {'neg_sample_size': args.neg_sample_size,
+                          'chunk_size': args.neg_sample_size,
+                          'chunk': True}
+        train_dataloader = KGETrainDataLoader(shuffle=args.shuffle_data,
+                                              batch_size=args.batch_size,
+                                              pin_memory=args.pin_memory,
```
Comment (classicsong) [inline python/refactor/main.py:68]: pin_memory can always be True.

File: python/refactor/main.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+                                              num_workers=args.num_workers,
+                                              drop_last=args.drop_last,
```
Comment (classicsong) [inline python/refactor/main.py:70]: drop_last can always be True.

File: python/refactor/main.py [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+                         'chunk_size': 1,
+                         'neg_sample_size': args.n_entities,
+                         'chunk': True}
+        eval_dataloader = KGEEvalDataLoader(batch_size=args.batch_size_eval,
+                                            pin_memory=args.pin_memory,
+                                            num_workers=args.num_workers,
+                                            metadata=eval_metadata)
+    else:
+        raise NotImplementedError(f'Dataloader for {args.model} is not supported yet.')
+    return train_dataloader, eval_dataloader
```
Comment (classicsong) [inline python/refactor/main.py:82]: Why there is no test_dataloader?

File: python/refactor/main.py [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+    args = TrainArgParser().parse_args()
+    prepare_args(args)
+
+    # print configuration
+    print('-' * 50)
+    for k, v in vars(args).items():
+        print('{} : {}'.format(k, v))
+    print('-' * 50)
+
+    set_seed(args.seed)
```
Comment (classicsong) [inline python/refactor/main.py:156]: We don't need this.

File: python/refactor/main.py [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+    init_time_start = time.time()
+
+    ############# create model #######################
+    model = create_model(args)
+    # setup logger
+    log_file = 'log.txt'
+    result_file = 'result.json'
```
Comment (classicsong) [inline python/refactor/main.py:163]: We can make it as default for Logger.

File: python/refactor/nn/metrics/metrics_evaluator.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+class MetricsEvaluator:
+    def __init__(self):
+        pass
+
+    def evaluate(self, results, data, graph):
+        # different models might have different results. Thus results might contain different number
+        # of elements to evaluate performance
+        pass
+
+class KGEMetricsEvaluator(MetricsEvaluator):
```
Comment (classicsong) [inline python/refactor/nn/metrics/metrics_evaluator.py:16]: It is better to call it RankingMetricsEvaluator. The method is not KGE specific.

File: python/refactor/nn/modules/decoder.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+    def load(self, load_path: str):
+        for decoder in self.decoder.items():
+            decoder.load(load_path)
+
+    def share_memory(self):
+        for decoder in self.decoder.items():
+            decoder.share_memory()
+
+    def prepare_model(self, gpu_id, rank=0, world_size=-1):
```
Comment (classicsong) [inline python/refactor/nn/modules/decoder.py:35]: We can remove this. As we are not considering DDP.

File: python/refactor/nn/modules/decoder.py [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+    def decode(self, encoded_data, data, idx=None):
+        ret_val = []
+        for name, decoder in self.decoder.items():
+            idx = self.decoder_index[name]
+            ret_val += [decoder.decode(encoded_data, data, idx)]
+        return ret_val
+
+    def evaluate(self, results: list):
+        for result, decoder in zip(results, self.decoder.values()):
+            decoder.decode(result)
```
Comment (classicsong) [inline python/refactor/nn/modules/decoder.py:54]: Decoder seems not define the decode(result) function.

File: python/refactor/nn/modules/decoder.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+        self.decoder = {}
+        self.decoder_index = {}
+
+    def add_module(self, name, module):
+        self.decoder[name] = module
+        self.decoder_index[name] = len(self.idx)
+
+    def save(self, save_path: str):
+        for decoder in self.decoder.items():
+            decoder.save(save_path)
```
Comment (classicsong) [inline python/refactor/nn/modules/decoder.py:25]: We need to define save, load, ... funcs in Decoder.

File: python/refactor/nn/modules/embedding.py [hunk 12]
```diff
@@ -0,0 +1,67 @@
+import torch as th
+from torch.nn import Embedding
+import numpy as np
+import copy
+import os
+
+
+class KGEmbedding:
```
Comment (classicsong) [inline python/refactor/nn/modules/embedding.py:8]: We need to use dgl.NodeEmbedding. I will handle it.

File: python/refactor/nn/modules/encoder.py [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+                 n_relation,
+                 init_func,
+                 rel_parts=None, # ! rel_parts and cross_rels must be put here otherwise harm the general structure
+                 cross_rels=None,
+                 soft_rel_part=False,
+                 strict_rel_part=False,
+                 encoder_name='KGEEncoder',):
+        super(KGEEncoder, self).__init__(encoder_name)
+        self.entity_emb = KGEmbedding()
+        self.relation_emb = KGEmbedding()
```
Comment (classicsong) [inline python/refactor/nn/modules/encoder.py:29]: We can use torch nn embedding here.

File: python/refactor/nn/modules/encoder.py [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+        self.async_update = args.async_update
+        self.async_threads = args.num_thread
+        self.pin_memory = args.pin_memory
+
+    def set_test_params(self, args):
+        self.pin_memory = args.pin_memory
+
+    def prepare_model(self, gpu_id, rank, world_size):
+        if not self.init:
+            if self.async_update:
```
Comment (classicsong) [inline python/refactor/nn/modules/encoder.py:88]: In the first version, we do not need async update.

File: python/refactor/nn/modules/model.py [hunk 15]
```diff
@@ -0,0 +1,359 @@
...
+from itertools import chain
+from tqdm import trange, tqdm
+from utils import get_scalar
+from .module import Module
+from utils.multiprocess import thread_wrapped_func
+import time
+
+# We use template design pattern here - which is divide the whole code chunk into partitions
+# so that user can define their own method if necessary.
+class Model(Module):
```
Comment (classicsong) [inline python/refactor/nn/modules/model.py:12]: Add some docstring here.

File: python/refactor/nn/modules/model.py [hunk 16]
```diff
@@ -0,0 +1,359 @@
...
+             profile=False):
+        self.eval()
+        start = time.time()
+        if world_size > 1:
+            self.share_memory()
+
+        if num_test_proc > 1:
+            queue = mp.Queue(num_test_proc)
+            procs = []
+            for i in range(num_test_proc):
```
Comment (classicsong) [inline python/refactor/nn/modules/model.py:80]: No multi-process for the first version

File: python/refactor/nn/modules/score_fun.py [hunk 17]
```diff
@@ -639,3 +647,141 @@ def fn(heads, relations, tails, num_chunks, chunk_size, neg_sample_size):
...
+    def load(self, path, name):
+        file_name = os.path.join(path, name)
+        # TODO: lingfei - determine whether to map location here
+        self.load_state_dict(th.load(file_name))
+
+    def create_neg(self, neg_head):
+        pass
+
+
+class ATTHScore(nn.Module):
```
Comment (classicsong) [inline python/refactor/nn/modules/score_fun.py:775]: What is ATTHScore?

File: python/refactor/utils/data/sampler.py [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from torch.utils.data import Sampler
+import numpy as np
+import torch as th
+
+
+class SequentialEpochSampler(Sampler):
```
Comment (classicsong) [inline python/refactor/utils/data/sampler.py:24]: Can you add some docstring here?

**Review by classicsong — COMMENTED** [review]
Some comments need to be addressed

File: python/refactor/main.py (hunk 2 above)
Comment (zheng-da) [reply python/refactor/main.py:39]: is this only used for KGE?

File: python/refactor/main.py [hunk 19]
```diff
@@ -0,0 +1,208 @@
...
+                             cross_rels=args.cross_rels,
+                             soft_rel_part=args.soft_rel_part,
+                             strict_rel_part=args.strict_rel_part)
+        return encoder
+    else:
+        raise NotImplementedError(f'encoder {args.encoder} is not supported yet.')
+
+def create_decoder(args):
+    if args.decoder == 'KGE':
+        decoder = KGEDecoder(args.decoder)
```
Comment (zheng-da) [inline python/refactor/main.py:109]: in concept, what does a decoder do? how is it different from a score function?

File: python/refactor/nn/loss/loss_criterion/__init__.py [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+softplus = functional.softplus
+sigmoid = th.sigmoid
+
+
+class BaseLoss(object):
+    def __call__(self, score, label):
+        raise NotImplementedError
+
+
+class HingeLoss(BaseLoss):
```
Comment (zheng-da) [inline python/refactor/nn/loss/loss_criterion/__init__.py:14]: why do we need to implement these loss functions? does pytorch not have them?

File: python/refactor/nn/modules/decoder.py [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+        pass
+
+    def set_test_params(self, args):
+        pass
+
+    def sync_model(self, gpu_id, rank, world_size):
+        pass
+
+    def postprocess_model(self, gpu_id, rank, world_size):
+        pass
```
Comment (zheng-da) [inline python/refactor/nn/modules/decoder.py:94]: why are the functions above specific to the KGE decoder? they shouldn't be used by other decoders?

File: python/refactor/nn/modules/encoder.py [hunk 22]
```diff
@@ -0,0 +1,138 @@
...
+        self.strict_rel_part = strict_rel_part
+        self.init = False
+
+    def save(self, save_path):
+        self.entity_emb.save(save_path, 'entity_emb.npy')
+        self.relation_emb.save(save_path, 'relation_emb.npy')
+
+    def load(self, load_path):
+        self.entity_emb.load(load_path, 'entity_emb.npy')
+        self.relation_emb.load(load_path, 'relation_emb.npy')
```
Comment (zheng-da) [inline python/refactor/nn/modules/encoder.py:52]: as we discussed last time, we should move relation_emb to KGE decoder, right?

File: python/refactor/nn/modules/encoder.py [hunk 23]
```diff
@@ -0,0 +1,138 @@
...
+                 hidden_dim,
+                 lr,
+                 optimizer,
+                 n_entity,
+                 n_relation,
+                 init_func,
+                 rel_parts=None, # ! rel_parts and cross_rels must be put here otherwise harm the general structure
+                 cross_rels=None,
+                 soft_rel_part=False,
+                 strict_rel_part=False,
```
Comment (zheng-da) [inline python/refactor/nn/modules/encoder.py:25]: we'll move most of the arguments as the arguments of the fit function.

File: python/refactor/nn/modules/module.py [hunk 24]
```diff
@@ -0,0 +1,77 @@
+from refactor.nn.loss.loss_generator import LossGenerator
+from refactor.nn.metrics import MetricsEvaluator
+from collections import OrderedDict
+
+# encoder, decoder are both inherited from Module. The default behavior of get_loss(), evaluate()
+# does nothing as sometimes encoder does not need them. While for decoder, it's mandatory for it
+# to overload these two functions.
+
+
+class Module(object):
```
Comment (zheng-da) [inline python/refactor/nn/modules/module.py:10]: are we trying to not rely on Pytorch?

File: python/refactor/utils/data/dataloader/dataloader.py [hunk 25]
```diff
@@ -0,0 +1,169 @@
+import torch as th
+from torch.utils.data.dataloader import DataLoader
+
+class CustomizeDataLoader:
```
Comment (zheng-da) [inline python/refactor/utils/data/dataloader/dataloader.py:4]: the name is confusing. it seems to be used to create pytorch dataloader. it doesn't load data from files either. can we call it DataLoaderGenerator?

File: python/refactor/utils/data/dataloader/dataloader.py [hunk 26]
```diff
@@ -0,0 +1,169 @@
...
+                 worker_init_fn=None,
+                 multiprocessing_context=None,
+                 generator=None,
+                 prefetch_factor=2,
+                 persistent_workers=False,
+                 num_workers=0):
+        self.batch_size = batch_size
+        self.shuffle = shuffle
+        self.sampler = sampler
+        self.batch_sampler = batch_sampler
```
Comment (zheng-da) [inline python/refactor/utils/data/dataloader/dataloader.py:22]: what are the two arguments?

File: python/refactor/utils/data/dataloader/dataloader.py [hunk 27]
```diff
@@ -0,0 +1,169 @@
...
+        self.pin_memory = pin_memory
+        self.drop_last = drop_last
+        self.timeout = timeout
+        self.worker_init_fn = worker_init_fn
```
Comment (zheng-da) [inline python/refactor/utils/data/dataloader/dataloader.py:26]: why do we need this?

File: python/refactor/utils/data/dataloader/dataloader.py [hunk 28]
```diff
@@ -0,0 +1,169 @@
...
+                 prefetch_factor=2,
+                 persistent_workers=False,
+                 num_workers=0):
+        self.batch_size = batch_size
+        self.shuffle = shuffle
+        self.sampler = sampler
+        self.batch_sampler = batch_sampler
+        self.pin_memory = pin_memory
+        self.drop_last = drop_last
+        self.timeout = timeout
```
Comment (zheng-da) [inline python/refactor/utils/data/dataloader/dataloader.py:25]: why do we need this?

File: python/refactor/utils/data/dataloader/dataloader.py [hunk 29]
```diff
@@ -0,0 +1,169 @@
...
+        self.worker_init_fn = worker_init_fn
+        self.multiprocessing_context = multiprocessing_context
+        self.generator = generator
+        self.prefetch_factor = prefetch_factor
+        self.persistent_workers = persistent_workers
```
Comment (zheng-da) [inline python/refactor/utils/data/dataloader/dataloader.py:30]: what are the arguments above?

File: python/refactor/utils/data/dataloader/util.py [hunk 30]
```diff
@@ -0,0 +1,27 @@
+# we use aggregation to aggregate sample data into DataWrapper. So that encoder&decoder can
+# have different behavior with these args defined.
+# one benifit of using DataWrapper instead of passing additional arguments directly to encoder&decoder
+# is that, we do not need to provide encoder(), decode() with other args that is different from case to case
+# so that the encode, decode method is eazier to extend.
+# we use adapter Design Pattern here.
```
Comment (zheng-da) [inline python/refactor/utils/data/dataloader/util.py:6]: i don't understand the comment above.

File: python/refactor/main.py (hunk 5 above)
Comment (menjarleev) [reply python/refactor/main.py:82]: Both validation and test follow the same logic that it uses same number of process, validation batch size, negative sample size, etc. while train might take different parameters settings such as chunk training, etc. During the process, the eval_dataloader can take either validation or test dataset and generates corresponding sub data sampler for each process.

File: python/refactor/nn/loss/loss_criterion/__init__.py (hunk 20 above)
Comment (menjarleev) [reply python/refactor/nn/loss/loss_criterion/__init__.py:14]: This is something of old version, we can remove it if we want.

File: python/refactor/nn/modules/encoder.py (hunk 22 above)
Comment (menjarleev) [reply python/refactor/nn/modules/encoder.py:52]: I believed we said that node_emb and relation_emb should be in encoder, and other stuff in decoder. I will double check with Xiang.

File: python/refactor/nn/modules/module.py (hunk 24 above)
Comment (menjarleev) [reply python/refactor/nn/modules/module.py:10]: Resolved, I've converted it into nn.Module

File: python/refactor/nn/modules/encoder.py (hunk 23 above)
Comment (menjarleev) [reply python/refactor/nn/modules/encoder.py:25]: I delete most of the arguments as it's for multi-GPU.

File: python/refactor/nn/modules/decoder.py (hunk 21 above)
Comment (menjarleev) [reply python/refactor/nn/modules/decoder.py:94]: deleted, it's for multi-GPU.
